Some Extractors fail with ConcurrentModificationException

I have tested the following extractors till now (it does not matter if they are together in a pipeline or not, results are same):
- RedirectExtractor
- PageIdExtractor
- PageLinksExtractor
- LabelExtractor
- ArticlePageExtractor

Out of the above, only RedirectExtractor and PageIdExtractor work properly, and the resulting outputs match with with the original extraction-framework. Adding any of the other 3 into the pipeline causes this:

```
...
...

May 22, 2014 9:17:17 PM org.dbpedia.extraction.dump.extract.ExtractionProgress start
INFO: li: 1 extractors (LabelExtractor), 1 datasets (labels) started
14/05/22 21:17:17 ERROR actor.OneForOneStrategy: java.util.ConcurrentModificationException
Serialization trace:
classes (sun.misc.Launcher$AppClassLoader)
classLoader (akka.actor.ReflectiveDynamicAccess)
_pm (akka.actor.ActorSystemImpl)
actorSystem (org.apache.spark.storage.BlockManager)
blockManager (org.apache.spark.storage.BlockManagerSource)
blockManagerSource (org.apache.spark.SparkContext)
_sparkContext (org.dbpedia.extraction.dump.extract.DistConfigLoader$$anon$1)
context (org.dbpedia.extraction.mappings.LabelExtractor)
extractors (org.dbpedia.extraction.mappings.CompositeParseExtractor)
extractor (org.dbpedia.extraction.mappings.RootExtractor)
extractor (org.dbpedia.extraction.dump.extract.ExtractorMapper)
com.esotericsoftware.kryo.KryoException: java.util.ConcurrentModificationException
Serialization trace:
classes (sun.misc.Launcher$AppClassLoader)
classLoader (akka.actor.ReflectiveDynamicAccess)
_pm (akka.actor.ActorSystemImpl)
actorSystem (org.apache.spark.storage.BlockManager)
blockManager (org.apache.spark.storage.BlockManagerSource)
blockManagerSource (org.apache.spark.SparkContext)
_sparkContext (org.dbpedia.extraction.dump.extract.DistConfigLoader$$anon$1)
context (org.dbpedia.extraction.mappings.LabelExtractor)
extractors (org.dbpedia.extraction.mappings.CompositeParseExtractor)
extractor (org.dbpedia.extraction.mappings.RootExtractor)
extractor (org.dbpedia.extraction.dump.extract.ExtractorMapper)
    at com.esotericsoftware.kryo.serializers.FieldSerializer$ObjectField.write(FieldSerializer.java:585)
    at com.esotericsoftware.kryo.serializers.FieldSerializer.write(FieldSerializer.java:213)
    at com.esotericsoftware.kryo.Kryo.writeObject(Kryo.java:501)
    at com.esotericsoftware.kryo.serializers.FieldSerializer$ObjectField.write(FieldSerializer.java:564)
    at com.esotericsoftware.kryo.serializers.FieldSerializer.write(FieldSerializer.java:213)
    at com.esotericsoftware.kryo.Kryo.writeObject(Kryo.java:501)
    at com.esotericsoftware.kryo.serializers.FieldSerializer$ObjectField.write(FieldSerializer.java:564)
    at com.esotericsoftware.kryo.serializers.FieldSerializer.write(FieldSerializer.java:213)
    at com.esotericsoftware.kryo.Kryo.writeObject(Kryo.java:501)
    at com.esotericsoftware.kryo.serializers.FieldSerializer$ObjectField.write(FieldSerializer.java:564)
    at com.esotericsoftware.kryo.serializers.FieldSerializer.write(FieldSerializer.java:213)
    at com.esotericsoftware.kryo.Kryo.writeObject(Kryo.java:501)
    at com.esotericsoftware.kryo.serializers.FieldSerializer$ObjectField.write(FieldSerializer.java:564)
    at com.esotericsoftware.kryo.serializers.FieldSerializer.write(FieldSerializer.java:213)
    at com.esotericsoftware.kryo.Kryo.writeObject(Kryo.java:501)
    at com.esotericsoftware.kryo.serializers.FieldSerializer$ObjectField.write(FieldSerializer.java:564)
    at com.esotericsoftware.kryo.serializers.FieldSerializer.write(FieldSerializer.java:213)
    at com.esotericsoftware.kryo.Kryo.writeObject(Kryo.java:501)
    at com.esotericsoftware.kryo.serializers.FieldSerializer$ObjectField.write(FieldSerializer.java:564)
    at com.esotericsoftware.kryo.serializers.FieldSerializer.write(FieldSerializer.java:213)
    at com.esotericsoftware.kryo.Kryo.writeObject(Kryo.java:501)
    at com.esotericsoftware.kryo.serializers.FieldSerializer$ObjectField.write(FieldSerializer.java:564)
    at com.esotericsoftware.kryo.serializers.FieldSerializer.write(FieldSerializer.java:213)
    at com.esotericsoftware.kryo.Kryo.writeClassAndObject(Kryo.java:568)
    at com.twitter.chill.TraversableSerializer$$anonfun$write$1.apply(Traversable.scala:30)
    at com.twitter.chill.TraversableSerializer$$anonfun$write$1.apply(Traversable.scala:28)
    at scala.collection.immutable.List.foreach(List.scala:318)
    at com.twitter.chill.TraversableSerializer.write(Traversable.scala:28)
    at com.twitter.chill.TraversableSerializer.write(Traversable.scala:21)
    at com.esotericsoftware.kryo.Kryo.writeObject(Kryo.java:501)
    at com.esotericsoftware.kryo.serializers.FieldSerializer$ObjectField.write(FieldSerializer.java:564)
    at com.esotericsoftware.kryo.serializers.FieldSerializer.write(FieldSerializer.java:213)
    at com.esotericsoftware.kryo.Kryo.writeObject(Kryo.java:501)
    at com.esotericsoftware.kryo.serializers.FieldSerializer$ObjectField.write(FieldSerializer.java:564)
    at com.esotericsoftware.kryo.serializers.FieldSerializer.write(FieldSerializer.java:213)
    at com.esotericsoftware.kryo.Kryo.writeObject(Kryo.java:501)
    at com.esotericsoftware.kryo.serializers.FieldSerializer$ObjectField.write(FieldSerializer.java:564)
    at com.esotericsoftware.kryo.serializers.FieldSerializer.write(FieldSerializer.java:213)
    at com.esotericsoftware.kryo.Kryo.writeClassAndObject(Kryo.java:568)
    at org.apache.spark.serializer.KryoSerializerInstance.serialize(KryoSerializer.scala:124)
    at org.dbpedia.extraction.spark.serialize.KryoSerializer$.serialize(KryoSerializer.scala:25)
    at org.dbpedia.extraction.spark.serialize.KryoSerializationWrapper.getValueSerialized(KryoSerializationWrapper.scala:19)
    at org.dbpedia.extraction.spark.serialize.KryoSerializationWrapper.writeObject(KryoSerializationWrapper.scala:32)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:606)
    at java.io.ObjectStreamClass.invokeWriteObject(ObjectStreamClass.java:988)
    at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1493)
    at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1429)
    at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1175)
    at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1541)
    at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1506)
    at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1429)
    at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1175)
    at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1541)
    at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1506)
    at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1429)
    at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1175)
    at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:347)
    at org.apache.spark.serializer.JavaSerializationStream.writeObject(JavaSerializer.scala:28)
    at org.apache.spark.scheduler.ResultTask$.serializeInfo(ResultTask.scala:48)
    at org.apache.spark.scheduler.ResultTask.writeExternal(ResultTask.scala:123)
    at java.io.ObjectOutputStream.writeExternalData(ObjectOutputStream.java:1456)
    at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1427)
    at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1175)
    at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:347)
    at org.apache.spark.serializer.JavaSerializationStream.writeObject(JavaSerializer.scala:28)
    at org.apache.spark.serializer.JavaSerializerInstance.serialize(JavaSerializer.scala:48)
    at org.apache.spark.scheduler.DAGScheduler.org$apache$spark$scheduler$DAGScheduler$$submitMissingTasks(DAGScheduler.scala:778)
    at org.apache.spark.scheduler.DAGScheduler.org$apache$spark$scheduler$DAGScheduler$$submitStage(DAGScheduler.scala:724)
    at org.apache.spark.scheduler.DAGScheduler.processEvent(DAGScheduler.scala:554)
    at org.apache.spark.scheduler.DAGScheduler$$anonfun$start$1$$anon$2$$anonfun$receive$1.applyOrElse(DAGScheduler.scala:190)
    at akka.actor.ActorCell.receiveMessage(ActorCell.scala:498)
    at akka.actor.ActorCell.invoke(ActorCell.scala:456)
    at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:237)
    at akka.dispatch.Mailbox.run(Mailbox.scala:219)
    at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(AbstractDispatcher.scala:386)
    at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
    at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
    at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
    at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
Caused by: java.util.ConcurrentModificationException
    at java.util.Vector$Itr.checkForComodification(Vector.java:1156)
    at java.util.Vector$Itr.next(Vector.java:1133)
    at com.esotericsoftware.kryo.serializers.CollectionSerializer.write(CollectionSerializer.java:67)
    at com.esotericsoftware.kryo.serializers.CollectionSerializer.write(CollectionSerializer.java:18)
    at com.esotericsoftware.kryo.Kryo.writeObject(Kryo.java:501)
    at com.esotericsoftware.kryo.serializers.FieldSerializer$ObjectField.write(FieldSerializer.java:564)
    ... 81 more
```

--- a/distributed/src/main/scala/org/dbpedia/extraction/dump/extract/DistConfigLoader.scala
+++ b/distributed/src/main/scala/org/dbpedia/extraction/dump/extract/DistConfigLoader.scala
@@ -6,26 +6,18 @@ import org.dbpedia.extraction.destinations.DeduplicatingDestination
 import org.dbpedia.extraction.destinations.Destination
 import org.dbpedia.extraction.destinations.MarkerDestination
 import org.dbpedia.extraction.destinations.WriterDestination
-import org.dbpedia.extraction.mappings.CompositeParseExtractor
-import org.dbpedia.extraction.mappings.Disambiguations
-import org.dbpedia.extraction.mappings.Extractor
-import org.dbpedia.extraction.mappings.Mappings
-import org.dbpedia.extraction.mappings.MappingsLoader
-import org.dbpedia.extraction.mappings.Redirects
-import org.dbpedia.extraction.mappings.RootExtractor
+import org.dbpedia.extraction.mappings._
 import org.dbpedia.extraction.ontology.io.OntologyReader
-import org.dbpedia.extraction.sources.{WikiPage, XMLSource, WikiSource}
+import org.dbpedia.extraction.sources.{Source, WikiPage, XMLSource, WikiSource}
 import org.dbpedia.extraction.util._
-import org.dbpedia.extraction.dump.download.Download
 import org.dbpedia.extraction.util.RichFile.wrapFile
 import org.dbpedia.extraction.wikiparser.Namespace
 import scala.collection.mutable.{ArrayBuffer, HashMap}
 import java.io._
 import java.net.URL
 import java.util.logging.{Level, Logger}
-import org.apache.spark.{SparkContext, SparkConf}
-import org.dbpedia.extraction.mappings.DistRedirects
 import org.apache.spark.rdd.RDD
+import org.dbpedia.extraction.dump.download.Download
 
 /**
  * Loads the dump extraction configuration.
@@ -34,10 +26,13 @@ import org.apache.spark.rdd.RDD
  *
  * TODO: get rid of all config file parsers, use Spring
  * TODO: Inherit ConfigLoader methods and get rid of redundant code
+ *
+ * @param config General extraction framework configuration
+ * @param distConfig Distributed configuration
  */
 class DistConfigLoader(config: Config, distConfig: DistConfig) extends ConfigLoader(config)
 {
-  private val logger = Logger.getLogger(classOf[ConfigLoader].getName)
+  private val logger = Logger.getLogger(classOf[DistConfigLoader].getName)
 
   /**
    * Loads the configuration and creates extraction jobs for all configured languages.
@@ -60,8 +55,51 @@ class DistConfigLoader(config: Config, distConfig: DistConfig) extends ConfigLoa
 
     val date = latestDate(finder)
 
-    //Extraction Context
-    val context = new DumpExtractionContext
+    val _articlesSource =
+    {
+      val articlesReaders = readers(config.source, finder, date)
+
+      XMLSource.fromReaders(articlesReaders, lang,
+                            title => title.namespace == Namespace.Main || title.namespace == Namespace.File ||
+                              title.namespace == Namespace.Category || title.namespace == Namespace.Template)
+    }
+
+    SparkUtils.silenceSpark()
+    val sparkContext = SparkUtils.getSparkContext(distConfig)
+
+    val cache = finder.file(date, "articles-rdd")
+
+    // Getting the WikiPages from local on-disk cache saves processing time.
+    val articlesRDD: RDD[WikiPage] = try
+    {
+      logger.info("Loading articles from cache file " + cache)
+      val loaded = DistIOUtils.loadRDD(sparkContext, classOf[WikiPage], cache.getAbsolutePath)
+      // count() throws org.apache.hadoop.mapred.InvalidInputException if file doesn't exist
+      val count = loaded.count()
+      logger.info(count + " WikiPages loaded from cache file " + cache)
+      loaded
+    }
+    catch
+      {
+        case ex: Exception =>
+        {
+          logger.log(Level.INFO, "Will read from wiki dump file for " + lang.wikiCode + " wiki, could not load cache file '" + cache + "': " + ex)
+          val articlesReaders = readers(config.source, finder, date)
+
+          val newRdd = sparkContext.parallelize(_articlesSource.toSeq, distConfig.sparkNumSlices)
+
+          DistIOUtils.saveRDD(newRdd, cache.getAbsolutePath)
+          newRdd
+        }
+      }
+
+    val _redirects =
+    {
+      val cache = finder.file(date, "template-redirects.obj")
+      DistRedirects.load(articlesRDD, cache, lang)
+    }
+
+    val contextBroadcast = sparkContext.broadcast(new DumpExtractionContext
     {
       def ontology = _ontology
 
@@ -95,67 +133,7 @@ class DistConfigLoader(config: Config, distConfig: DistConfig) extends ConfigLoa
 
       def mappings: Mappings = _mappings
 
-      private val _sparkContext =
-      {
-        SparkUtils.silenceSpark()
-        SparkUtils.getSparkContext(distConfig)
-      }
-
-      def sparkContext: SparkContext = _sparkContext
-
-      //This isn't needed here, until we make it choosable - distributed (with Spark) vs single-node extraction
-      private lazy val _articlesSource =
-      {
-        val articlesReaders = readers(config.source, finder, date)
-
-        XMLSource.fromReaders(articlesReaders, language,
-                              title => title.namespace == Namespace.Main || title.namespace == Namespace.File ||
-                                title.namespace == Namespace.Category || title.namespace == Namespace.Template)
-      }
-
-      def articlesSource = _articlesSource
-
-      private val _articlesRDD =
-      {
-        val cache = finder.file(date, "articles-rdd")
-
-        // Getting the WikiPages from local on-disk cache saves processing time.
-        val rdd: RDD[WikiPage] = try
-        {
-          logger.info("Loading articles from cache file " + cache)
-          val loaded = DistIOUtils.loadRDD(sparkContext, classOf[WikiPage], cache.getAbsolutePath)
-          // count() throws org.apache.hadoop.mapred.InvalidInputException if file doesn't exist
-          val count = loaded.count()
-          logger.info(count + " WikiPages loaded from cache file " + cache)
-          loaded
-        }
-        catch
-          {
-            case ex: Exception =>
-            {
-              logger.log(Level.INFO, "Will read from wiki dump file for " + lang.wikiCode + " wiki, could not load cache file '" + cache + "': " + ex)
-              val articlesReaders = readers(config.source, finder, date)
-
-              val newRdd = sparkContext.parallelize(XMLSource.fromReaders(articlesReaders, language,
-                                                                   title => title.namespace == Namespace.Main || title.namespace == Namespace.File ||
-                                                                     title.namespace == Namespace.Category || title.namespace == Namespace.Template).toSeq, distConfig.sparkNumSlices)
-
-              DistIOUtils.saveRDD(newRdd, cache.getAbsolutePath)
-              newRdd
-            }
-          }
-
-        rdd
-      }
-
-      def articlesRDD = _articlesRDD
-
-      private val _redirects =
-      {
-        val cache = finder.file(date, "template-redirects.obj")
-        val redirects = DistRedirects.load(articlesRDD, cache, language)
-        redirects
-      }
+      def articlesSource: Source = _articlesSource
 
       def redirects: Redirects = _redirects
 
@@ -174,7 +152,10 @@ class DistConfigLoader(config: Config, distConfig: DistConfig) extends ConfigLoa
       }
 
       def disambiguations: Disambiguations = if (_disambiguations != null) _disambiguations else new Disambiguations(Set[Long]())
-    }
+
+    })
+
+    val context = new DistDumpExtractionContext(contextBroadcast)
 
     //Extractors
     val extractor = CompositeParseExtractor.load(extractorClasses, context)
@@ -197,10 +178,45 @@ class DistConfigLoader(config: Config, distConfig: DistConfig) extends ConfigLoa
     val destination = new MarkerDestination(new CompositeDestination(formatDestinations.toSeq: _*), finder.file(date, Extraction.Complete), false)
 
     val description = lang.wikiCode + ": " + extractorClasses.size + " extractors (" + extractorClasses.map(_.getSimpleName).mkString(",") + "), " + datasets.size + " datasets (" + datasets.mkString(",") + ")"
-    new DistExtractionJob(new RootExtractor(extractor), context.articlesRDD, config.namespaces, destination, lang.wikiCode, description)
+    new DistExtractionJob(new RootExtractor(extractor), articlesRDD, config.namespaces, destination, lang.wikiCode, description)
   }
 
   // TODO: Inherit methods below from ConfigLoader?
+
+  //language-independent val
+  private val _ontology =
+  {
+    val ontologySource = if (config.ontologyFile != null && config.ontologyFile.isFile)
+    {
+      XMLSource.fromFile(config.ontologyFile, Language.Mappings)
+    }
+    else
+    {
+      val namespaces = Set(Namespace.OntologyClass, Namespace.OntologyProperty)
+      val url = new URL(Language.Mappings.apiUri)
+      val language = Language.Mappings
+      WikiSource.fromNamespaces(namespaces, url, language)
+    }
+
+    new OntologyReader().read(ontologySource)
+  }
+
+  //language-independent val
+  private val _commonsSource =
+  {
+    try
+    {
+      val finder = new Finder[File](config.dumpDir, Language("commons"), config.wikiName)
+      val date = latestDate(finder)
+      XMLSource.fromReaders(readers(config.source, finder, date), Language.Commons, _.namespace == Namespace.File)
+    } catch
+      {
+        case ex: Exception =>
+          logger.info("Could not load disambiguations - error: " + ex.getMessage)
+          null
+      }
+  }
+
 
   private def writer(file: File): () => Writer =
   {
@@ -229,32 +245,6 @@ class DistConfigLoader(config: Config, distConfig: DistConfig) extends ConfigLoa
     logger.info(s"Source is ${source} - ${files.size} file(s) matched")
 
     files
-  }
-
-  //language-independent val
-  private lazy val _ontology =
-  {
-    val ontologySource = if (config.ontologyFile != null && config.ontologyFile.isFile)
-    {
-      XMLSource.fromFile(config.ontologyFile, Language.Mappings)
-    }
-    else
-    {
-      val namespaces = Set(Namespace.OntologyClass, Namespace.OntologyProperty)
-      val url = new URL(Language.Mappings.apiUri)
-      val language = Language.Mappings
-      WikiSource.fromNamespaces(namespaces, url, language)
-    }
-
-    new OntologyReader().read(ontologySource)
-  }
-
-  //language-independent val
-  private lazy val _commonsSource =
-  {
-    val finder = new Finder[File](config.dumpDir, Language("commons"), config.wikiName)
-    val date = latestDate(finder)
-    XMLSource.fromReaders(readers(config.source, finder, date), Language.Commons, _.namespace == Namespace.File)
   }
 
   private def latestDate(finder: Finder[_]): String =

--- a/distributed/src/main/scala/org/dbpedia/extraction/dump/extract/DistDumpExtractionContext.scala
+++ b/distributed/src/main/scala/org/dbpedia/extraction/dump/extract/DistDumpExtractionContext.scala
@@ -1,0 +1,31 @@
+package org.dbpedia.extraction.dump.extract
+
+import org.dbpedia.extraction.util._
+import org.dbpedia.extraction.mappings._
+import org.dbpedia.extraction.sources._
+import org.dbpedia.extraction.ontology.Ontology
+import org.apache.spark.broadcast.Broadcast
+
+/**
+ * Wrapper for DumpExtractionContext to be used with Spark.
+ *
+ * Delegates all methods to a DumpExtractionContext broadcast variable.
+ */
+class DistDumpExtractionContext(contextBroadcast: Broadcast[_ <: DumpExtractionContext]) extends DumpExtractionContext
+{
+  override def ontology: Ontology = contextBroadcast.value.ontology
+
+  override def commonsSource: Source = contextBroadcast.value.commonsSource
+
+  override def language: Language = contextBroadcast.value.language
+
+  override def mappingPageSource: Traversable[WikiPage] = contextBroadcast.value.mappingPageSource
+
+  override def mappings: Mappings = contextBroadcast.value.mappings
+
+  override def articlesSource: Source = contextBroadcast.value.articlesSource
+
+  override def redirects: Redirects = contextBroadcast.value.redirects
+
+  override def disambiguations: Disambiguations = contextBroadcast.value.disambiguations
+}

